sdl/storage.go

```diff
@@ -3,18 +3,143 @@ package sdl
 import (
 	"sort"
 
+	"github.com/pkg/errors"
 	"gopkg.in/yaml.v3"
 
 	"github.com/ovrclk/akash/types"
 )
 
+const (
+	storageAttributePersistence = string("persistence")
+	storageAttributeClass       = string("class")
+	storageAttributeMount       = string("mount")
+	storageAttributeReadOnly    = string("readOnly") // we might not need it at this point of time
+)
+
+var (
+	errUnsupportedStorageAttribute = errors.New("sdl: unsupported storage attribute")
+	errStorageMountPoint           = errors.New("sdl: persistent storage must have mount point")
+	errStorageMultipleEphemeral    = errors.New("sdl: multiple ephemeral storages are not allowed")
+)
+
 type v2StorageAttributes types.Attributes
 
 type v2ResourceStorage struct {
 	Quantity   byteQuantity        `yaml:"size"`
 	Attributes v2StorageAttributes `yaml:"attributes,omitempty"`
 }
 
+type v2ResourceStorageArray []v2ResourceStorage
+
+var allowedStorageAttributes = map[string]bool{
+	storageAttributePersistence: true,
+	storageAttributeClass:       true,
+	storageAttributeMount:       true,
+	storageAttributeReadOnly:    true,
+}
+
+// UnmarshalYAML unmarshal storage config
+// data can be present either as single entry mapping or an array of them
+// e.g
+// single entity
+// ```yaml
+// storage:
+//   quantity: 1Gi
+//   attributes:
+//     class: ssd
+// ```
+//
+// ```yaml
+// storage:
+//   - quantity: 512Mi # ephemeral storage
+//   - quantity: 1Gi
+//     attributes:
+//       class: ssd
+//   - quantity: 100Gi
+//     attributes:
+//       persistent: true # this volumes survives pod restart
+//       class: gp # aka general purpose
+// ```
+func (sdl *v2ResourceStorageArray) UnmarshalYAML(node *yaml.Node) error {
+	var nodes v2ResourceStorageArray
+
+	switch node.Kind {
+	case yaml.SequenceNode:
+		for _, content := range node.Content {
+			var nd v2ResourceStorage
+			if err := content.Decode(&nd); err != nil {
+				return err
+			}
+
+			nodes = append(nodes, nd)
+		}
+	case yaml.MappingNode:
+		var nd v2ResourceStorage
+		if err := node.Decode(&nd); err != nil {
+			return err
+		}
+
+		nodes = append(nodes, nd)
+	}
+
+	// check storage parameters in SDL are set properly
+	// 1. there can be only one ephemeral entry
+	//    ephemeral considered as one without mount point
+	// 2. ephemeral cannot have persistence
+	// 3. ephemeral cannot be readOnly (for now. configMaps are considered as ephemeral as well)
+	// todo make this as helper
+	ephemeral := false
+	for _, nd := range nodes {
+		attr := make(map[string]string)
+
+		for _, nd := range nd.Attributes {
+			attr[nd.Key] = nd.Value
```

Comment:
This would be easier to read as 

```
attrPersistence := false
attrMount := false

for k, v := range nd.Attributes {
  if k == storageAttributePersistence {
    attrPersistence = len(v.Value) ! = 0 // Convert to bool
} else if k == storageAttributeMount {
   attrMount = len(v.Value) != 0 // Convert to bool
}
```

Then you can just use the variables below.